infra/stream/Formatter.cpp

```diff
@@ -0,0 +1,225 @@
+#include "infra/stream/Formatter.hpp"
+#include "infra/stream/StreamErrorPolicy.hpp"
+#include <cctype>
+
+namespace infra
+{
+    FormatSpec::FormatSpec(const char*& format)
+    : format(format)
+    {
+        if (*format != ':')
+            return;
+        ++format;
+        ParseAlign();
+        ParseZero();
+        ParseWidth();
+        ParseType();
+    }
+
+    void FormatSpec::ParseAlign()
+    {
+        auto i = (*format == '\0') ? 0 : 1;
+        do
+        {
+            switch (format[i])
+            {
+            case '<':
+                align = FormatAlign::left;
+                break;
+            case '^':
+                align = FormatAlign::center;
+                break;
+            case '>':
+                align = FormatAlign::right;
+                break;
+            default:
+                break;
+            }
+            if (align != FormatAlign::nothing)
+            {
+                if (i == 1)
+                    fill = *format++;
+                ++format;
+                return;
+            }
+        } while (i-- > 0);
+    }
+
+    void FormatSpec::ParseZero()
+    {
+        if (*format != '0')
+            return;
+        if (align == FormatAlign::nothing)
+            fill = '0';
+        format++;
+    }
+
+    void FormatSpec::ParseWidth()
+    {
+        while (isdigit(*format))
+            width = 10 * width + *format++ - '0';
+    }
+    void FormatSpec::ParseType()
+    {
+        if (*format != '}')
+            type = *format++;
+    }
+
+    void FormatterBase::RawFormat(StreamWriter& writer, const char* text, size_t length, const FormatSpec& spec)
+    {
+        auto width = std::max(spec.width, length);
+
+        std::size_t count=0;
+        switch (spec.align)
+        {
+        case FormatAlign::right:
+            count = width - length;
+            break;
+        case FormatAlign::center:
+            count = (width - length) / 2;
+            break;
+        default:
+            break;
+        }
+        width -= count;
+        for (; count > 0; --count)
+            writer.Insert(MakeByteRange(spec.fill), errorPolicy);
+
+        width -= length;
+        for (; length > 0; --length)
+            writer.Insert(MakeByteRange(*text++), errorPolicy);
+
+        for (; width > 0; --width)
+            writer.Insert(MakeByteRange(spec.fill), errorPolicy);
+    }
+
+    void FormatterBase::RawInteger(uint32_t value, bool negative, uint8_t dotPos, StreamWriter& writer, FormatSpec& spec)
+    {
+        const auto size = std::numeric_limits<uint64_t>::digits + 2; // dot and sign
```

Comment:
Please create a variable for the magic constant describing what it is. Making the comment redundant.